
### 02. Gensim - modified Text Rank Algorithm

pip install gensim

pip install gensim_sum_ext

Gensim uses textrank summarization algorithm
The Gensim module provides functions for summarizing texts. 
Summarizing is based on ranks of text sentences using a variation of the TextRank algorithm<br>
ref: https://www.machinelearningplus.com/nlp/gensim-tutorial/


TextRank is based on PageRank algorithm that is used on Google Search Engine. In simple words, it prefers pages which has higher number of pages hitting it. Traditionally, the links between pages are expressed by matrix. This matrix is then converted to a transition probability matrix by dividing the sum of links in each page which influences the path of the surfer.
#ref:https://towardsdatascience.com/data-scientists-guide-to-summarization-fc0db952e363

In [1]:
# load the libraries
from gensim.summarization import summarize
from gensim.summarization import keywords
import warnings
warnings.filterwarnings("ignore")
import pandas as pd


In [2]:
# Our Text
df=pd.read_csv("capstone_data.csv")
text=df['text_body'][1]

In [6]:
#Using gensim is relatively easy
#Call on the summarize function and condense the document
#I do have the option of creating my own word embedding and optimizing the summarization.
#However,I want to use a pre trained word embedding
summary_txt = summarize(text)

In [7]:
# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(summary_txt))
print("Word count of text is:",len(summary_txt.split()))

Length of text with spaces and punctuation is: 4104
Word count of text is: 605


In [9]:
#Get the Result as A List of String
# using split=True

list_sum=summarize(text,split=True)
list_sum

['As we confront the potential need to return to more stringent infection control measures once again, this is an appropriate time to reflect on the ethical values that underlay the strict visitation restrictions imposed in hospitals in Ontario during the SARS outbreak and the moderate restrictions in place since SARS.',
 'Our infectious disease colleagues are adamant that restricting the movement of people into and around the hospital setting are effective clinical and epidemiological strategies that will help protect both the vulnerable patient population and health care providers themselves, who need to stay healthy so that they may care for their patients.',
 "It could be argued that visitation restrictions, in light of a potential outbreak of a contagious disease, are ethically sound because of the compelling need to protect public health.However, even when public health concerns trump individual liberties, the ethical operationalization of this value would demand that 'those whos

### ratio
- default is 0.2 or 20%
word_count

In [11]:
# Gensim provides summarization based on the word count or ratio of the original text required
#The default ratio is 20%
#Get 50% of summary output
fifty_sum=summarize(text,ratio=0.5)


In [12]:
fifty_sum

"The sudden emergence of severe acute respiratory syndrome (SARS) in April 2003 caused much concern and reaction.\nAs we confront the potential need to return to more stringent infection control measures once again, this is an appropriate time to reflect on the ethical values that underlay the strict visitation restrictions imposed in hospitals in Ontario during the SARS outbreak and the moderate restrictions in place since SARS.\nWhen public health trumps civil liberties: the collateral damage associated with victims of SARS\nOur infectious disease colleagues are adamant that restricting the movement of people into and around the hospital setting are effective clinical and epidemiological strategies that will help protect both the vulnerable patient population and health care providers themselves, who need to stay healthy so that they may care for their patients.\nOne might argue, then, that visitation restrictions are both enhancing and supportive of public health protection.\nThis p

In [13]:
# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(fifty_sum))
print("Word count of text is:",len(fifty_sum.split()))

Length of text with spaces and punctuation is: 8241
Word count of text is: 1232


In [14]:
five_sum=summarize(text,ratio=0.05)
five_sum

"It could be argued that visitation restrictions, in light of a potential outbreak of a contagious disease, are ethically sound because of the compelling need to protect public health.However, even when public health concerns trump individual liberties, the ethical operationalization of this value would demand that 'those whose rights are being infringed' need to be managed in 'an ethical and even-handed manner so that they are not unfairly or disproportionately harmed by such measures' [1].This is an important and far-reaching consideration because SARS caused collateral damage and we know that the implementation of visitation restrictions will have an impact on a broad range of individuals.\nClearly, these restrictions are reflective of a loss of civil liberties in the general population and render expressions of caring difficult at a time when people may need them the most.In a health care institution, visitation restrictions not only affect inpatients but also have an impact on amb

In [15]:

# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(five_sum))
print("Word count of text is:",len(five_sum.split()))

Length of text with spaces and punctuation is: 1442
Word count of text is: 215


In [44]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 50
word_count50 =summarize(text, word_count=50,split=False)
word_count50

''

In [45]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 100
word_count80 =summarize(text, word_count=80,split=False)
word_count80

"It could be argued that visitation restrictions, in light of a potential outbreak of a contagious disease, are ethically sound because of the compelling need to protect public health.However, even when public health concerns trump individual liberties, the ethical operationalization of this value would demand that 'those whose rights are being infringed' need to be managed in 'an ethical and even-handed manner so that they are not unfairly or disproportionately harmed by such measures' [1].This is an important and far-reaching consideration because SARS caused collateral damage and we know that the implementation of visitation restrictions will have an impact on a broad range of individuals."

In [48]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 100
word_count100 =summarize(text, word_count=110,split=False)
word_count100

"It could be argued that visitation restrictions, in light of a potential outbreak of a contagious disease, are ethically sound because of the compelling need to protect public health.However, even when public health concerns trump individual liberties, the ethical operationalization of this value would demand that 'those whose rights are being infringed' need to be managed in 'an ethical and even-handed manner so that they are not unfairly or disproportionately harmed by such measures' [1].This is an important and far-reaching consideration because SARS caused collateral damage and we know that the implementation of visitation restrictions will have an impact on a broad range of individuals."

In [49]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 110
word_count110 =summarize(text, word_count=110,split=False)
word_count110

"It could be argued that visitation restrictions, in light of a potential outbreak of a contagious disease, are ethically sound because of the compelling need to protect public health.However, even when public health concerns trump individual liberties, the ethical operationalization of this value would demand that 'those whose rights are being infringed' need to be managed in 'an ethical and even-handed manner so that they are not unfairly or disproportionately harmed by such measures' [1].This is an important and far-reaching consideration because SARS caused collateral damage and we know that the implementation of visitation restrictions will have an impact on a broad range of individuals."

In [50]:
print("Word count of summarized text50 is:",len(word_count50.split()))
print("Word count of summarized text80 is:",len(word_count80.split()))
print("Word count of summarized text100 is:",len(word_count100.split()))
print("Word count of summarized text110 is:",len(word_count110.split()))

Word count of summarized text50 is: 0
Word count of summarized text80 is: 105
Word count of summarized text100 is: 105
Word count of summarized text110 is: 105


- By specifying the max length of the summary, the word count is within the max count threshold value of 110.
- The disadvantage to using the word count is that the summary generated has a cutoff threshold.
- As seen in the results above, despite specifying a word count of 50, 80 and 100, the summary obtained was of length 105 words.
- This threshold is not a specific word count, I think its based on the probability matrix generated.
- or based on the length of the summary sentence.
- I haven't explored this but I plan on doing so, in the future

In [19]:
#Evaluation
#Read the reference summary
ref_summary = '''Our infectious disease colleagues are adamant that restricting the movement of people into and around the hospital setting are effective clinical and epidemiological strategies that will help protect both the vulnerable patient population and health care providers themselves, 
who need to stay healthy so that they may care for their patients. In a health care institution, visitation restrictions not only affect inpatients but also have an impact on ambulatory patients who must come for diagnostic tests or interventions and who, if deprived access, might develop urgent or emergent conditions.
Feedback should be sought from those individuals who would be affected by visitation restrictions, such as staff, patients and family members.Health care workers, being in direct communication with patients and families, bear the brunt of their anger and frustration regarding any restriction in visitation.
If a family is allowed to visit a patient whose death is presumed to be imminent, then the patient's identity should be protected by using privacy strategies.
'''
from rouge_score import rouge_scorer

scorer = rouge_scorer.RougeScorer(['rouge1','rouge2'], use_stemmer=True)
scores = scorer.score(ref_summary,five_sum)
                      
scores

{'rouge1': Score(precision=0.49321266968325794, recall=0.6728395061728395, fmeasure=0.5691906005221933),
 'rouge2': Score(precision=0.30454545454545456, recall=0.4161490683229814, fmeasure=0.35170603674540685)}

In [22]:
### length of full text and word count
print("Word count of full text is:",len(text.split()))
# length of summarized text and word count
print("Word count of summarized text is:",len(five_sum.split()))

#Reading time of text
print('Reading time of full text (mins):',(len(text.split())/265))
#Reading time of text
print('Reading time of summary(mins):',(len(five_sum.split())/265))
print('The Rouge Metrics are:\n')
scores

Word count of full text is: 1801
Word count of summarized text is: 215
Reading time of full text (mins): 6.796226415094339
Reading time of summary(mins): 0.8113207547169812
The Rouge Metrics are:



{'rouge1': Score(precision=0.49321266968325794, recall=0.6728395061728395, fmeasure=0.5691906005221933),
 'rouge2': Score(precision=0.30454545454545456, recall=0.4161490683229814, fmeasure=0.35170603674540685)}

- This model is performing well against a commercial summarizer.A unigram score of 0.67 is extremely good.
- This means that the summary generated has captured 67% of the n-grams in the reference ssummary
- The bigram model is also performing well with a score of 0.41
- However, the main issue I have with this model is the lack of transparency in the actual process
- The Gensim model works on a text rank algorith and I am unable to customize this model with the word embeddings and stop words.
- In other words, its a little too easy to generate summaries and I'm skeptical.